### PR TITLE
Run PREPARE on all nodes (Issue #106)

### PR DESCRIPTION
**Pull request by havaker**
This pull request implements #106 by sending PREPARE via all available connections.

### CODE REVIEW

**Comment by havaker** [comment]
@piodul @psarna please review

File: scylla/src/transport/session.rs [hunk 1]
```diff
@@ -148,17 +151,25 @@ impl Session {
...
-            _ => Err(anyhow!("Unexpected frame received")),
+        let connections = self.get_connections()?;
+        let all_connections = connections.iter().map(|(_, v)| v).flatten();
+
+        // Prepare statements on all connections concurrently
+        let handles = all_connections.map(|c| c.prepare(query));
+        let mut results = try_join_all(handles).await?;
+
+        // Check if every received statement has the same ID
+        for p in results.windows(2) {
```
Comment (psarna) [inline scylla/src/transport/session.rs:162]: Please move this additional check to a separate helper function (called validate_prepare_ids or something alike).
Also, consider writing it in a more declarative fashion, e.g.:
```rust
results.windows(2).all(|p| p[0].get_id() == p[1].get_id()) // + add error handling at the end of the chain
```

File: scylla/src/transport/session.rs [hunk 2]
```diff
@@ -148,17 +151,25 @@ impl Session {
...
-        match result {
-            Response::Error(err) => Err(err.into()),
-            Response::Result(result::Result::Prepared(p)) => Ok(PreparedStatement::new(
-                p.id,
-                p.prepared_metadata,
-                query.to_owned(),
-            )),
-            _ => Err(anyhow!("Unexpected frame received")),
+        let connections = self.get_connections()?;
+        let all_connections = connections.iter().map(|(_, v)| v).flatten();
```
Comment (piodul) [inline scylla/src/transport/session.rs:155]: Use `flat_map`:

```rust
let all_connections = connections.iter().flat_map(|(_, v)| v);
```

Comment (havaker) [reply scylla/src/transport/session.rs]: Done

File: scylla/src/transport/session.rs (hunk 1 above)
Comment (havaker) [reply scylla/src/transport/session.rs]: Done

**Comment by piodul** [comment]
@havaker There were two big PRs merged which caused this PR to have conflicts with the main branch - please rebase.

**Comment by havaker** [comment]
> @havaker There were two big PRs merged which caused this PR to have conflicts with the main branch - please rebase.

Done

File: scylla/src/transport/session.rs [hunk 3]
```diff
@@ -180,22 +183,29 @@ impl Session {
...
+        let mut results = try_join_all(handles).await?;
+
+        if !Self::validate_prepared_ids_equality(&results) {
+            return Err(TransportError::PreparedStatementsIDsNotMatch);
+        }
+
+        match results.pop() {
+            Some(result) => Ok(result),
+            None => Err(TransportError::NoConnectionsAvailable),
         }
```
Comment (piodul) [inline scylla/src/transport/session.rs:200]: Convert it to:
```rust
results.pop().ok_or_else(TransportError::NoConnectionsAvailable)
```

**Review by piodul — COMMENTED** [review]
Left one small nit, otherwise LGTM

File: scylla/src/transport/session.rs (hunk 3 above)
Comment (havaker) [reply scylla/src/transport/session.rs]: Done (I chose to use `ok_or` instead of `ok_or_else`)

**Review by psarna — APPROVED** [review]

**Review by piodul — APPROVED** [review]

**Pull request merged**